OpenWeatherMap Module

alexoneill opened the pull request:
Previous PR: #816 
> Thiss adds an alternate weather module, with a whole plethora of additional configuration options.
> 
> It is based on [OpenWeatherMap](https://openweathermap.org/), a free service that provides weather forecasting. It is also based on [IP API](http://ip-api.com/) for reverse location lookup based on IP.
> 
> The major goal of this module is to offer the user complete configuration for nearly any weather scenario, as well as be extensible and well-documented. Major features include:
> 
> - Automatic weather based on location
> - Formatting options for different weather data (rain, snow, barometric pressure, wind, etc.)
> - Support for multiple-day forecasts
> - Localization support, with a handful of popular languages supported.
> Let me know what you all think, and I'll happily make changes as you see fit. :)
> 
> I realize that this is a fairly large change request, so take you time to look it over. A good amount of it is documentation, and I tried to keep the source code as clean as possible.
> 
> Alex
> 

Had to make this a new PR since I changed the branch on my fork. Anyways, all of the above still holds true, plus a few more things:
- _No need for external modules_
- Similar configuration options to `weather_yahoo` so migration is easy
- More configuration for even _more_ customization
- Ability to have color-changing temperature based on a widely-used temperature-color mapping
- Reduced optimism ( @tobes )
- Made sure to only use Unicode in default values and verified compatibility with DejaVu Sans Mono

Here are some issues I would like some feedback on:
- ~~The linter seems to complain about my color options section, that those items are not present in the module documentation. Thoughts?~~
- I added an `os` call to read an API key from the command-line environment while testing to avoid committing an API key into the wild. Is this alright?
- I made some changes to non-module areas. Please take a look to make sure that these look alright!

As always, happy to take all suggestions for improvement!

Alex

tobes commented on py3status/request.py:
```diff
@@ -63,6 +63,7 @@ def __init__(self, url, params, data, headers, timeout, auth):
             elif isinstance(e, HTTPError):
                 self._status_code = e.code
                 self._error_message = reason
+                self._response = e
```
From what I can see this is unused.  Additionally it should not be being used by any module due to being named as 'private'

alexoneill replied:
Removed!

tobes commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,797 @@
+        self.py3.prevent_refresh()
+
+
+if (__name__ == '__main__'):
```
These parentheses are unneeded and should be removed this seems to be something you are doing quite a lot, it makes the code look like JavaScript :p

We should use them only where it clarifies things, here it just makes the code harder to read.

tobes commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,797 @@
+    color_pos_50 = '#7FFF00'
+    color_pos_70 = '#FF9900'
+    color_pos_90 = '#FF0000'
+    color_zero = '#0000FF'
```
These can probably be better handled using the thresholds functionality.  Also bear in mind that much of the world uses a different temperature scale.

alexoneill replied:
Happy to change these defaults. 

I used @lasers [suggestion](https://wattsupwiththat.com/2008/06/26/color-and-temperature-perception-is-everything/) and looked at the Weather Underground scale [here](https://wattsupwiththat.files.wordpress.com/2008/06/wunderground_national_temps.png).

I avoided using thresholds in favor of color mixing so the color is a continuous gradient between the different temperatures. If we want to not make it so fancy, I can always configure thresholds. :)

tobes replied:
>  avoided using thresholds in favor of color mixing so the color is a continuous gradient between the different temperatures. 

#766 add this functionality globally

alexoneill replied:
Great! I switched to thresholds! :)

tobes commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,797 @@
+   - {f}: Current Fahrenheit temperature reading.
+   - {k_max}: The maximum Kelvin temperature.
+   - {k_min}: The minimum Kelvin temperature.
+   - {k}: Current Kelvin temperature reading.
```
personally I'd do theses as `{temp}` `{temp_min}` etc and then allow the user to select a temp scale eg `F`, `C` or `K`

Also same with wind speed etc

alexoneill replied:
Fixed! I used some custom logic to verify that the unit passed to the configuration parameter is supported as well.

alexoneill commented on py3status/request.py:
```diff
@@ -63,6 +63,7 @@ def __init__(self, url, params, data, headers, timeout, auth):
             elif isinstance(e, HTTPError):
                 self._status_code = e.code
                 self._error_message = reason
+                self._response = e
```
I included this since I read the response when OWM gives a failure.

```python
req = self.py3.request(url, timeout=self.request_timeout)
if(req.status_code != 200):
    data = req.json()
    raise OWMException(data['message'])
```

According to the [docs](https://docs.python.org/2/library/urllib2.html#urllib2.HTTPError), the error object acts as a response-like object so you can interpret the content of the page.

The above would not work without the change here, since the `req.json()` would fail, since `self._response` would be undefined.

I can remove it, but it gives richer error messages to the end user. I could always assume that the error is API Key related, since that's usually the case, and it would remove the need for the change.

Thoughts?

tobes replied:
I'll look at this in more detail when I have time in a few days, thanks

tobes replied:
It looks like you are not using this any longer

alexoneill commented:
Hey all! Gentle ping here. What should be modified for this to be accepted?

Alex

lasers commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,708 @@
+        text = ''
+        if coords is not None:
+            wthr = self._get_weather(coords)
+            fcsts = self._get_forecast(coords)
```
Here, I think... We make double requests on every loop... only when we have the forecasts enabled.
Should we try and see if we can obtain both `wthr` and `fcsts` with one `_make_req()` instead?

ultrabug replied:
👍 indeed, is it possible ?

alexoneill replied:
The main issue here is that the forecast output does not include sunrise / sunset time, as far as I can tell.

The documentation is a little sparse, but there is no indication that these two things are included in the forecast output. Unfortunately, that info is in the current forecast information, hence the need.

alexoneill replied:
For reference, the relevant documentation is here:

- [Current Forecast Docs](https://openweathermap.org/forecast5)
- [Future Forecast Docs](https://openweathermap.org/current)

ultrabug commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,708 @@
+            self.forecast_text_separator, forecasts)
+        return self.py3.safe_format(self.format, today)
+
+    def weather(self):
```
yes please

alexoneill replied:
Fixed!

ultrabug commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,708 @@
+            'icon': self._get_icon(wthr),
+            'clouds': self._format_clouds(wthr),
+            'rain': self._format_rain(wthr),
+            'snow': self._format_rain(wthr),
```
good catch lasers!

alexoneill replied:
Awesome catch! Fixed! :)

ultrabug commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,708 @@
+    forecast_days = 0
+    forecast_include_today = False
+    forecast_text_separator = ' '
+    format = '{icon}: {temp} {desc}'
```
I agree, explicit is better please

alexoneill replied:
Fixed! If I missed any, let me know!

ultrabug commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,708 @@
+# -*- coding: utf-8 -*-
+'''
+Ultimately customizable weather module based on the IP-API Geolocation API
```
This can be ignored for this PR imo. If it becomes a problem, we will address it later.

alexoneill replied:
I added it anyways, it wasn't difficult!

ultrabug commented:
Thanks for this huge work @alexoneill, minor things still as lasers pointed out and then we'll proceed 👍

ultrabug commented:
Ok @alexoneill your recent changes look good thanks for this.

I have now tested your module and I've got a couple or remarks from a user perspective, I'd like you to tell me if my testing is dumb or not.

I started with the basic, added the module with no config

- got the error on the bar, fair enough

I added a valid api_key parameter

- nothing showed up, empty result = no module display

**I would expect the default config to display something, don't you ?**

I tried to set `forecast_days` to 1

- nothing showed up

**Here I am confused ?**

I configured `forecast_days` to 2

- I got only one result displayed

Am I missing something mate ?

Cheers

alexoneill commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,708 @@
+        (default 'mph')
+
+Format Placeholders:
+    - All:
```
Fixed!

alexoneill commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,708 @@
+        # Generate our icon array
+        self.icons = self._get_icons()
+
+        # Conversion factors
```
Fixed!

alexoneill commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,708 @@
+        self.kmh_from_msec = 0.277778
+        self.mph_from_msec = 2.23694
+
+        # Verify the units configuration
```
Fixed!

alexoneill commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,708 @@
+            'cached_until': self.py3.time_in(seconds=self.cache_timeout)
+        }
+
+    def on_click(self, event):
```
Fixed!

alexoneill commented:
@ultrabug 

I updated the default configuration to be a bit more user friendly. I added forecast information and increased the default number of for the forecast from zero to three.

Hmm. About the blank output: I suspect what's happening is that your instance is generating a request error when connecting to the IP-API for determining your location. When this happens, the module falls back to displaying nothing.

This differs from the behavior when the module cannot contact OWM (the weather service), where it throws a user-visible error. Do you think this behavior should be changed?

[Here's a video](https://asciinema.org/a/zI9hO9Ph54L07ZqrdSGm2N68x) of what I see. The default configuration is the following:
```python
import os
from py3status.module_test import module_test

module_test(Py3status, config={
    'api_key': os.getenv('OWM_API_KEY'),
})
```
I first ran the module with the default config, then disabled my wifi, ran it again, enabled the wifi, and ran it one final time.

Any ideas why it's failing for you? I want to make sure we can fix this so everyone can use the module :)

lasers commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,739 @@
+        },
+
+        # Complete configuration
+        'format_cloud': '{icon}: {coverage}%',
```
Super happy to have `{location}` placeholder. It worked great. Things changed when I played with `(lat, long)`. Still uneasiness because we don't have `{city.name}` placeholder. Would be ideal for laptops on the go... and possibly good to have it part of the default format too. This module is getting bigger and bigger. :o

And this is a typo. We might want to rename the others to `format_cloud` to prevent little things like this. I also see `clouds` for `icons`. Maybe it should be `cloud` too.

alexoneill replied:
Fixed the `format_cloud` part! I'll add a city name option in a separate commit!

alexoneill replied:
Added the city output! :) Unfortunately, it is only available in `format` and not `format_forecast`, but that should be fine.

tobes commented on py3status/modules/weather_owm.py:
```diff
@@ -31,7 +31,7 @@
         (regardless of the 'forecast_include_today' flag)
-        (default 0)
+        (default 3)
     forecast_include_today: Include today in the forecast? (Boolean)
```
I think they are both still needed

alexoneill replied:
I agree with @tobes, the end user may want to have the current weather in both.

Also, this option is for better cross compatibility with `weather_yahoo`, since it also has this option.

tobes commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,739 @@
+    format_humidity = '{icon}: {humidity}%'
+    format_pressure = '{icon}: {pressure} hPa'
+    format_rain = '{icon}: {amount} inches'
+    format_snow = '{icon}: {amount} inches'
```
> Is it because we have 0 for {amount}

This shouldn't affect things

alexoneill replied:
Ah! Thanks for catching that. Fixed!

ultrabug commented:
@alexoneill Thanks a lot for your patience and your work.

The default is good for me now!

> This differs from the behavior when the module cannot contact OWM (the weather service), where it throws a user-visible error. Do you think this behavior should be changed?

Well TBH I'm not sure that being offline should make the module disappear if it succeeded to get the weather once. It's not like the weather changes so much in a few.

=> Maybe you could change the color to `self.py3.COLOR_DEGRADED` when you display stale results because you're offline (or the API is down) ?

Apart from that, I guess lasers and tobes outlined some quick stuff to fix as well.

Almost there mate, cheers!

ultrabug commented:
PS: I discovered asciicinema thx to you, love it 👍

alexoneill commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,741 @@
+        (default 0)
+    forecast_include_today: Include today in the forecast? (Boolean)
+        (default False)
+    forecast_text_separator: Separator between entries in the forecast
```
This was left as `forecast_text_separator` to have better cross compatibility with `weather_yahoo`.

alexoneill commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,739 @@
+        Options:
+            cm, ft, in, mm, m, yrd
+        (default 'in')
+    temp_color: Color the temperature output based on a color scale.
```
I would like to leave the default in for `thresholds`, since it gives the user a reasonable temperature color scale built in.

I do like this idea, but I'm not sure how to leave a good default for the user and eliminate the `temp_color` option. 

It could be accomplished by special casing on the `thresholds` value, by having hard-coded, recognized options such as `thresholds = 'basic'`, and handling it from there, but that seems clunky.

I'm open to suggestions!

alexoneill commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,739 @@
+        defaults = {
+            'i200_i299': self.icon_thunderstorm,
+            'i300_i399': self.icon_rain,
+            'i500_i599': self.icon_rain,
```
Yeah, this is correct. These ranges of icons are for rain, and the extreme ones are correct as well. Thanks for checking! :)

alexoneill commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,739 @@
+
+        # Extract forecast
+        weathers = data['list']
+        return weathers[:-1] if (self.forecast_include_today) else weathers[1:]
```
See above discussion!

alexoneill commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,739 @@
+            'icon': self.icons['sunrise'],
+        })
+
+    def _format_sunset(self, wthr):
```
Hmm. The output here is in UTC. Let me see what I can do about it!

alexoneill replied:
I can get the current timezone information from the IP-API, but without an external Python dependency, I would be unable to convert UTC to the local timezone.

Fixing this is possible, and there is precedent to use an external timezone module (the `clock` module uses some), but it increases the barrier to entry so to speak.

Thoughts?

alexoneill replied:
I'm going to change my IP API provider, there is another one which should give me the proper information! :)

Look for another commit!

lasers replied:
Will do. (Friendly report: Time still wrong)

alexoneill commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,741 @@
+    format_temperature: Formatting for temperature
+        Available placeholders:
+            current, icon, max, min
+        (default '{icon}: {current}°')
```
Sure! I took out the colons, and I removed the comma.

alexoneill commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,739 @@
+       {min} The minimum temperature in the configured unit.
+       {current} Current temperature in the configured unit.
+    format_wind:
+       {degree} Current wind speed heading (in degrees).
```
I simplified descriptions! :)

alexoneill commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,739 @@
+    format_rain = '{icon}: {amount} inches'
+    format_snow = '{icon}: {amount} inches'
+    format_sunrise = '{icon}: %X'
+    format_sunset = '{icon}: %X'
```
Good suggestion, fixed!

alexoneill commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,739 @@
+    format_sunrise = '{icon}: %X'
+    format_sunset = '{icon}: %X'
+    format_temperature = u'{icon}: {current}°'
+    format_wind = '{icon}: {speed} mph'
```
Gotta love floating point! Fixed! :)

alexoneill commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,739 @@
+        return self.py3.safe_format(self.format_wind, choice)
+
+    def _format_humidity(self, wthr):
+        # Format the humidity (default zero humidity)
```
Added some logic to make sure an empty value gives an explicit zero!

lasers replied:
Still `●: %` here. Maybe the problem is not with zero.

alexoneill replied:
This should be fixed! :)

lasers commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,751 @@
+        'api_key': os.getenv('OWM_API_KEY'),
+
+        # Select icons
+        'icons': {
```
This does not work for me. Am I doing something wrong? This should replace `{icon}`  in `format_pressure`? 
This config tripped the `py3status`'s config checker and gave me red `CONFIG ERROR`. I guess it worked okay in terminal+testmode and not in real statusbar.
```
* Inside weather_owm... 

        icons {
            pressure = 'PRESSURE'
        }
```

alexoneill replied:
Fixed!

tobes commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,751 @@
+    format_snow: Formatting for snow volume over the past 3 hours
+        Available placeholders:
+            icon, amount
+        (default '{icon} {amount:.0f}in')
```
seeing how we know the unit `snow_unit` could we not us it here eg `'{icon} {amount:.0f}{snow_unit}'` so that when we change the unit the display updates too.

I'm not sure we need `snow_unit` and `rain_unit` etc maybe `unit_depth` or something but maybe the extra flexibility is worth it.

I still need to try this module for real but I think the idea behind it is great

ultrabug commented:
Hi guys, since I'd like this merged asap I think we should only concentrate on typos from now on.

Some of them have been highlighted by @lasers so please @alexoneill fix them and let's proceed.

ultrabug commented:
Friendly ping, I'd love that long work merged

alexoneill commented:
Hey all, sorry for the delay! Just getting back from vacation. I had limited network access, so I was unable to work on this, but I'll incorporate the above tonight! :)

Alex

alexoneill commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,751 @@
+        pressure['icon'] = self.icons['pressure']
+
+        # Format the barometric pressure
+        return self.py3.safe_format(self.format_humidity, pressure)
```
Fixed!

lasers commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,750 @@
+            'wind': 954,
+            'humidity': self.icon_humidity,
+            'pressure': self.icon_pressure,
+            'temperature': self.icon_temperature,
```
Something I'm wondering about... There could be an issue of figuring out static icons and dynamic icons. For instance, we don't have `icon_sunrise` + `icon_sunset` so where did they came from? Turns out that you assigned them with static icons: `sun` and `cloud` and if people want to change them, they should change them in `format_*`.... (only if there are no `icon_` to work with).

Would it be ideal to put all static icons in `format_*` instead? This kinda removes the "rule" of having `{icon}` in all `format_*`... and we also would have fewer `icon_*` options to parse. Did that make sense?
```
# Examples
    format_humidity = '● {humidity}%'
    format_pressure = '◌ {pressure}hPa'
    format_temperature = u'○ {current:.0f}°'
```
I'm not sure about other icons since they could be dynamic?...  just only three here? Maybe `sunset` and `sunrise` too... Just five? I can't be sure.

lasers replied:
Another confusing thing about customizing icons... is that I think it is possible to configure `icons = {xxx:x}` instead of `icon_xxx = x`. If that's accurate, then we technically could get rid of all `icon_xxx` options instead.

Having this split up possibly make things confusing. Easy to use `icon_xxx = x` over `icons = {xxx:x}`. Possibly easier (but ugly) if we don't split them up. Going with this route should reduce lines too... so I dunno what we should do here. Also, we only use `icons = {xxx:x}` to change just icons so it's not a terrible idea.

alexoneill replied:
I'm not quite sure what you mean by `icons = {xxx:x}`. Could you give a specific example?

I do agree that we could remove the `icon_xxx` options by statically placing them in the various `format_xxx` options!

tobes replied:
> icons = {xxx:x} 

please don't go here

lasers replied:
This works if the format have `{icon}` so maybe we need to keep them all supplied. Forget what I recently said about statically putting icons in various `format_xxx` if we were to to do this and/or make `icons = {}` fully workable. Example config.
```
        icons = {
            clouds:'CLOUD',
            humidity:'HUMIDITY',
            pressure:'PRESSURE',
            rain:'RAIN',
            snow:'SNOW',
            wind:'WIND',
            temperature: 'TEMPERATURE',
            sunset: 'SUNSET',
            sunrise: 'SUNRISE',
        }
```
Seeing how that's possible to customize icons using `icons = {xxx:x}` instead of outside `icon_xxx`... `icon_xxx` parts is now somewhat redundant. 

I don't know if @tobes / @ultrabug would like this idea. They shot down `button = {pause:1, stop:3, next:4, prev:5}` in new modules PRs. Your `icons = {xxx:x}` likely was made to work with range, but can simply work with standalone icons too (like what your desc said)... meaning we could live without all those `icon_xxx` config.

```
        Also, you can specify the icons for the various formatting sections
        below. For example, to specify the icon for 'format_pressure', use
        'pressure'. A few formatting sections will take from the defaults or be
        dynamic, including (clouds, rain, wind, sunrise, sunset).
        These may be specified regardless.
```

alexoneill replied:
I removed the ability to specify the general icon options (`cloud`, `rain`, `snow`, etc.) from the `icons` config option. This should reduce user confusion :)

alexoneill commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,750 @@
+    forecast_days = 3
+    forecast_include_today = False
+    forecast_text_separator = ' '
+    format = '{city}: {icon} {temperature} {main} {forecast}'
```
Hmm, not sure. I'm fine removing `{main}`, especially since it seems to mimic the longer form. Even then, `{description}` is never really that verbose, so maybe it makes sense...

alexoneill replied:
I've decided to leave both in, but update the default to the (slightly) more verbose `{description}` placeholder. Might as well give the user the option to have a slightly smaller output if they desire.

alexoneill commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,750 @@
+        if self.snow_unit not in SNOW_UNITS:
+            raise Exception('snow_unit is not recognized')
+        if self.temp_unit not in TEMP_UNITS:
+            raise Exception('temp_unit is not recognized')
```
Fixed! I let any capitalization work for units, so this should not be a problem :)

lasers commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,751 @@
+        'api_key': os.getenv('OWM_API_KEY'),
+
+        # Select icons
+        'icons': {
```
Please update `Example configuration` with `s/=/:` 👍

alexoneill replied:
Fixed!

lasers replied:
Please update `i666` to `666`. (Will look through things again). 👍

alexoneill replied:
Fixed! Thanks for the catch :)

lasers replied:
Please update `666` -> `'666'` and `666_999` -> `'666_999'`👍

ultrabug commented:
Welcome back around @alexoneill, hope you had a nice time :)

This ought to be the most commented PR ever, keep up the good work and thanks a lot for this great persistence in making a remarkable module 👍

alexoneill commented:
@ultrabug Haha, seems like its going that way! And not a problem. I love `py3status`, so I'm happy to contribute! :)

alexoneill commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,758 @@
+    temp_unit: Unit for temperature
+        Options:
+            c, f, k
+        (default 'f')
```
Fixed! Sorry, I should have updated `f` to `F` a few commits back, but it must have slipped my mind.

alexoneill commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,758 @@
+        The default below is intended for Fahrenheit. If the set value is empty
+        or None, the feature is disabled.
+        (default [(-100, '#0FF'), (0, '#00F'), (50, '#0F0'), (150, '#FF0')])
+    wind_unit: Unit for wind speed
```
Good call, it would certainly help with consistency since we have `icon_xxx` and `format_xxx`.

Fixed!

alexoneill commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,758 @@
+    format_humidity = '{icon} {humidity}%'
+    format_pressure = '{icon} {pressure} hPa'
+    format_rain = '{icon} {amount:.0f} {unit}'
+    format_snow = '{icon} {amount:.0f} {unit}'
```
Very cool, I had no idea about this feature. Added `{rain}` to the default config and added the conditional logic to `rain`, `snow`, and `wind`!

Plus I removed the colon on the city part, @lasers :)

alexoneill commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,758 @@
+                    if '_' in key:
+                        if key.count('_i') != 1:
+                            raise Exception('Icon range is not properly'
+                                            'formatted! (%s)' % key)
```
I cleaned it up a bit! :)

alexoneill commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,758 @@
+            'i801_i809': self.icon_cloud,
+            'i900_i909': self.icon_extreme,
+            'i950_i959': self.icon_wind,
+            'i960_i999': self.icon_extreme,
```
Fixed!

lasers commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,755 @@
+        Available placeholders:
+            icon, city, clouds, rain, snow, wind, humidity, pressure,
+            temperature, sunrise, sunset, main, description, forecast
+        (default '{city} {icon} {temperature}{rain}, {description} {forecast}')
```
I think I see what you're trying to do here. You want to undo doublespace in all `format_`.
```
[\?if=amount  {icon}... to...
[\?if=amount {icon}...
```
And use this format. 
```
format = '{city} {icon} {temperature}[ {rain},] {description} {forecast}'
```
If no `{rain}`, no space and comma. // P.S. Remove `format`'s parenthesis on `L291`.

alexoneill replied:
Fixed! Good call.

lasers replied:
(Fix me) `format_clouds` too. I'm currently at `1%` right now wondering if it'll reach `0%`.

lasers commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,755 @@
+        data = {}
+        for source in (defaults, self.icons):
+            for key in source:
+                if not re.match(r'[0-9]+', key):
```
(Opinion) We `import re` for one-time use here (only in `post_config_hook`). I think we could avoid `re` by using other options such as...
* if `key.is_digit()` # is `600` only.
* For `600_700`, something like.. if `len(key.split('_')) == 2 and key.replace('_', '').isdigit():`

alexoneill replied:
Fixed!

lasers commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,760 @@
+MPH_FROM_MSEC = 2.23694
+
+# Thresholds defaults
+THRESHOLDS = [
```
Would you like this thresholds? I think it's a bit more... **_COLORFUL_**! 🌈 
```
        THRESHOLDS = [
            (-65, '#111111'),
            (-60, '#21006B'),
            (-55, '#4C006B'),
            (-50, '#6B006B'),
            (-45, '#820082'),
            (-40, '#990099'),
            (-35, '#B300B3'),
            (-30, '#CC00CC'),
            (-25, '#E600E6'),
            (-20, '#FF00FF'),
            (-15, '#D100FF'),
            (-10, '#9E00FF'),
            (-5, '#6600FF'),
            (0, '#0000FF'),
            (5, '#004AFF'),
            (10, '#0073FF'),
            (15, '#00A3FF'),
            (20, '#00CCFF'),
            (25, '#00E6FF'),
            (30, '#00FFFF'),
            (35, '#00FFB3'),
            (40, '#7FFF00'),
            (45, '#CEFF00'),
            (50, '#FFFF00'),
            (55, '#FFE600'),
            (60, '#FFCC00'),
            (65, '#FFAE00'),
            (70, '#FF9900'),
            (75, '#FF7F00'),
            (80, '#FF4F00'),
            (85, '#FF0000'),
            (90, '#FF4545'),
            (95, '#FF6868'),
            (100, '#FF8787'),
            (105, '#FF9E9E'),
            (110, '#FFB5B5'),
            (115, '#FFCFCF'),
            (120, '#FFE8E8'),
            (125, '#EEEEEE')
        ]
````

tobes replied:
this is why we have #766

tobes replied:
we should consider users of different units 30C is pretty warm.

lasers replied:
Slightly different issue, but relevant.... I'm stuck with this coloring issue.

We can deal with `thresholds` in `K` unit so we can obtain the color hex correctly before we translate the unit... And instead of coloring the `format_temperature`, we colorize the temperature value itself... that way, we can give `max` and `min` temperatures their own color.

That works. We should get: `%orange%`70`%nocolor%`°F here.

**Issue**: If users want to colorize other things? They can't.

How do we allow users to colorize their `{unit}` and user-input characters in `format_*`? Do we have or need something like a `thresholds` formatter?

* `[\?thresholds=current {current:0.f}°{unit}]` should convert to...
  `[\?color=#FF4F00 {current:0.f}°{unit}]`.

Otherwise, the author of this module could... for `x` in `{current}`, `{max}`, `{min}`... do merge colorized `{x}°{unit}` together into `{x}` and users won't be able to remove `°F` if they wanted to, but still colorized... I did this in `coin_market` too... with `+` and `%`.

tobes replied:
> Do we have or need something like a thresholds formatter?

this is actually done via the color stuff.  thresholds can have names that then get set as the color via `self.py3.threshold(value, name)` and then we can do `\?color=name` or if no name `\?color=threshold`

tobes commented on tests/test_module_doc.py:
```diff
@@ -46,6 +46,7 @@
     ('kdeconnector', '_dev'),  # move to __init__ etc
     ('arch_updates', 'format'),  # dynamic
     ('spotify', 'sanitize_words'),  # line too long for docstring parsing
+    ('weather_owm', 'icons'),  # complicated
```
I think this is no longer needed too

alexoneill replied:
Fixed!

alexoneill commented:
Hey all! 

I fixed the above issues. Let me know what else should change! You can check out my recent commits, but nothing too big was changed. Take a look at the threshold logic though, that's where the bulk of the editing went.

Sorry for the off-and-on contributions recently, I've had low bandwidth these past few weeks, but the holiday weekend has given me some time to contribute! :)

lasers commented:
Add `{country}` too for lulz.

alexoneill commented on tests/test_module_doc.py:
```diff
@@ -46,6 +46,7 @@
     ('kdeconnector', '_dev'),  # move to __init__ etc
     ('arch_updates', 'format'),  # dynamic
     ('spotify', 'sanitize_words'),  # line too long for docstring parsing
+    ('weather_owm', 'thresholds'),  # dictionary parsing issue
```
I couldn't get the default here to play nicely with the test suite. Thoughts?

tobes commented:
@alexoneill cool.

regarding the thresholds and the tests, for now this is fine I think the thing is to get the functionality how you want and then we can sort any small details.

Are you now happy with this module?  If so I'll do a new review of it.  If you still want to play some more that's good too.  Just let me know either way.

alexoneill commented:
@lasers I added the country!

@tobes Yes, I am certainly happy with this now, please review! I will only make bug fixes from here on out :)

lasers commented on py3status/modules/weather_owm.py:
```diff
@@ -0,0 +1,811 @@
+            * To have smooth transitions between colors, consider setting the
+              'gradients' configuration parameter to 'True', either in the
+              global configuration, or in the module configuration!
+        (default {'all': [(-100, '#0FF'), (0, '#00F'), (50, '#0F0'), (150, '#FF0')]})
```
This trips on a list.
`thresholds = [(-100, '#0FF'), (0, '#00F'), (50, '#0F0'), (150, '#FF0')]`
```
AttributeError: 'list' object has no attribute 'keys'
  File "/home/chris/src/py3status/py3status/module.py", line 132, in prepare_module
    self.module_class.post_config_hook()
  File "/home/chris/src/py3status/py3status/modules/weather_owm.py", line 417, in post_config_hook
    if set(self.thresholds.keys()) > THRESHOLDS_NAMES:
2017-09-10 17:04:40 INFO terminating module weather_owm
```
@tobes Maybe you want to review this part because you're more familiar with thresholds? This is making things either easier or complicated when it's going to play favorites with the temperatures.

tobes replied:
I'll take a look, but a quick skim makes me think this is not a problem we probably just need to remove this bit of code anyway.

alexoneill replied:
Friendly ping here, what should I do to help out? :)

tobes replied:
Sorry things have been a bit busy the last few weeks

ultrabug commented:
I am happy to merge this long ongoing work!

Thanks a lot for your patience and great contribution @alexoneill :)

alexoneill commented:
@ultrabug @lasers @tobes, everyone,

Thank you very much for your feedback and ultimately your acceptance into py3status! I'm super happy that this can help others with weather information, and I'm looking forward to making more PRs.

This project is incredible, I use it every day. Thank you for putting in the hard work and dedication to maintain it for the rest of us. :)

Alex
